### PR TITLE
Added first attempt of __repr__ for TensorMap and TensorBlock

### PR DESCRIPTION
**Pull request by DavideTisi**
In the spirit of issue #32. I added an idea for two `__repr__` functions for TensorMap and TensorBlock.
In TensorMap it basically prints the keys and its names:
```
>> Tensor
TensorMap 
keys: [(0, 1, 1), (0, 1, 6), (0, 1, 8), (0, 6, 1), (0, 6, 6), (0, 6, 8), (0, 8, 1), (0, 8, 6), (0, 8, 8)]
key_names: ['spherical_harmonics_l', 'species_center', 'species_neighbor']
```
For the TensorBock, the function prints the names of the samples, components and properties:

```
>> Tensor.block(0)
TensorBlock 
samples: ['structure', 'center']
component: ['spherical_harmonics_m']
properties: ['n']
```

The example are taken from a spherical harmonic expansion Tensor.

### CODE REVIEW

**Comment by Luthaf** [comment]
The build failure is unrelated, and should be fixed on master. Can you rebase this branch to bring in the fix?

Some small suggestions for the output
```bash
TensorMap with 9 blocks  # add the number of blocks
# how about a vertical layout? takes more space but might be easier to read
keys: ['spherical_harmonics_l', 'species_center', 'species_neighbor'] 
               0,                        1,               1,
               0,                        1,               6,

                                        ... # skip some keys if they are more than 10?
                                        
               0,                        8,               6,
               0,                        8,               8,
```

If you agree with this, maybe we can skip keys only in `__repr__` and not in `__print__` (similarly to how numpy prints arrays)
---

For the tensor block, I would also add some information about the gradients

**Comment by DavideTisi** [comment]
I agree for sure with the addition of the number of blocks.
the vertical layout how you would suggest aligning the numbers with just tabs?
I mean if the keys_names changes it would be difficult are different to put the number under the label.

Regarding the skip, I agree to put it only in __repr__ (actually that's why I haven't put it, normally I do not like when it skips stuff.

I can complete the TensorBlock with gradients. Here do we put full infos or just if there are gradients or not? and then a __repr__ (or similar) function for the gradient only?

**Comment by Luthaf** [comment]
>  the vertical layout how you would suggest aligning the numbers with just tabs?
> I mean if the keys_names changes it would be difficult are different to put the number under the label.

There is some code doing this kind of alignment in rust, here: https://github.com/lab-cosmo/equistore/blob/8823476d864409618afb23c96dc86affc767b160/src/labels.rs#L229-L244

It is not too bad: you can use the length of the different variables as padding when writing the values. I think python formatting can take the padding width as an extra parameter (not sure what's the syntax for it though).

> I can complete the TensorBlock with gradients. Here do we put full infos or just if there are gradients or not? and then a repr (or similar) function for the gradient only?

I think just having which gradients are present or not is fine, and then having another `__repr__` function in the `Gradient` proxy object is a great idea!

**Comment by DavideTisi** [comment]
The TensorMap prints the number of blocks and the values in a table-like way.
The padding for sure can be improved (also the code is not super good-looking).
 
If the number of blocks is < 10, both functions do not skip any value.
```
>> Tensor
TensorMap with 9 blocks 
keys: ['spherical_harmonics_l' 'species_center' 'species_neighbor]
                  0                   1                1
                  0                   1                6
                  0                   1                8
                  0                   6                1
                  0                   6                6
                  0                   6                8
                  0                   8                1
                  0                   8                6
                  0                   8                8

>> print(Tensor)
TensorMap with 9 blocks 
keys: ['spherical_harmonics_l' 'species_center' 'species_neighbor]
                  0                   1                1
                  0                   1                6
                  0                   1                8
                  0                   6                1
                  0                   6                6
                  0                   6                8
                  0                   8                1
                  0                   8                6
                  0                   8                8
```

If the number of blocks is > 10.


```
>> Tensor
TensorMap with 63 blocks 
keys: ['spherical_harmonics_l' 'species_center' 'species_neighbor]
                  0                   1                1
                  1                   1                1
                  2                   1                1
               ...
                  4                   8                8
                  5                   8                8
                  6                   8                8
```

`print` function never skip.


The TensorBlock output now gives info on the gradients

If there are no gradients:
```
TensorBlock 
samples: ['structure', 'center']
component: ['spherical_harmonics_m']
properties: ['n']
gradients: No
```

If there are gradients, it gives the list of parameters wrt the gradients are computed:

```
>> Tensor.block(0)
TensorBlock 
samples: ['structure', 'center']
component: ['spherical_harmonics_m']
properties: ['n']
gradients: ['positions']
```

Finally the gradients block:
```
>> Tensor.block(0).gradient('positions')
Gradient TensorBlock 
parameter: 'positions'
samples: ['sample', 'structure', 'atom']
component: ['gradient_direction', 'spherical_harmonics_m']
properties: ['n']
```

**Comment by Luthaf** [comment]
I like the new output! 

The build will fail again, you'll need to reformat the code with [black](https://github.com/psf/black). You can do this manually with `tox -e format` or integrate it automatically in your editor with https://github.com/microsoft/vscode-black-formatter (I know you are using vscode). The automatic integration will automatically format all python files when you save them.

**Review by Luthaf — COMMENTED** [review]
A couple of small suggestions, but the output looks fine to me. Could you also add some basic tests?

File: tests/python/blocks.py [hunk 1]
```diff
@@ -1,11 +1,25 @@
...
+            values=np.full((3, 2), -1.0),
+            samples=Labels(["samples"], np.array([[0], [2], [4]], dtype=np.int32)),
+            components=[],
+            properties=Labels(["properties"], np.array([[5], [3]], dtype=np.int32)),
+        )
+        ROOT = os.path.dirname(__file__)
+        tfile = os.path.join(ROOT, "..", "repr_string_block.data")
+        with open(tfile, "r") as f:
+            s = f.readlines()
+        self.assertTrue(block.__repr__() == "".join(s))
```
[the remark this answers is not in the document]
Comment (DavideTisi) [reply tests/python/blocks.py:21]: you're right

File: tests/python/utils.py [hunk 2]
```diff
@@ -3,10 +3,12 @@
 from equistore import Labels, TensorBlock, TensorMap
 
 
-def test_tensor_map():
+def test_tensor_map(iflong: bool = False):
```
Comment (Luthaf) [inline tests/python/utils.py:6]: You can remove the `iflong` parameter now =)

File: tests/python/tensor.py [hunk 3]
```diff
@@ -1,19 +1,40 @@
...
+    def test_print(self):
+        """
+        Test routine for the print function of the TensorBlock.
+        It compare the reults with those in a file.
+        """
+        tensor = test_tensor_map()
+        repr = tensor.__repr__()
+        ROOT = os.path.dirname(__file__)
+        tfile = os.path.join(ROOT, "..")
+        with open(tfile + "/repr_string.data", "r") as f:
```
Comment (Luthaf) [inline tests/python/tensor.py:28]: could you move the data out of the files and directly inside the code here as well?

**Pull request merged**